### ISSUE TITLE
Native BFloat16 Mixed BatchNorm Train gives incorrect gradients

### ISSUE THREAD

**Issue by dnikolaev-amd** [pytorch/pytorch#156513]
### 🐛 Describe the bug

`Native NCHW BFloat16 Mixed Train` gives incorrect gradients (for weight, bias, input) comparing with both MIOpen and CPU (and Native NHWC). MIOpen and CPU give the same result. 

test file `test_batchnorm_native.py`
```python
import contextlib
import torch
from torch.testing._internal.common_cuda import _get_torch_rocm_version
from torch.testing._internal.common_utils import TestCase, instantiate_parametrized_tests,  parametrize, run_tests

class TestBatchNormNative(TestCase):
    def _assertEqualWrapper(self, a, b, msg="") -> int:
        """      
        Helper function to compare two tensors and print a message if they are not equal.
        Allows to count and print all errors (not only the first one).
        :param a: first tensor
        :param b: second tensor
        :param msg: message to print if tensors are not equal
        :return: 0 if tensors are equal, 1 if they are not."""
        try:
            self.assertEqual(a, b)
            return 0
        except Exception as e:
            print(f"\n{msg} >>> {e}")
            return 1

    @parametrize("test_backend", ['native', 'miopen'], name_fn= lambda x: f'backend_{x}')
    def test_batchnorm(self, test_backend):

        if test_backend == 'miopen' and (not torch.version.hip or _get_torch_rocm_version() < (6,4)):
            self.skipTest("bfloat16 mixed batchnorm is supported only on ROCm 6.4 and later")

        size = (4, 8, 2, 2)

        # create reference inputs and model on CPU
        ref_input = torch.randn(size=size, dtype=torch.bfloat16, device='cpu').requires_grad_()
        ref_grad = torch.randn(size=size, dtype=torch.bfloat16, device='cpu')
        ref_mod = torch.nn.BatchNorm2d(size[1], device='cpu', dtype=torch.float) #mixed
        ref_mod.weight.data.uniform_()
        ref_mod.bias.data.uniform_()
        
        # create inputs and model on GPU (native or miopen)
        test_input = ref_input.detach().clone().to(device='cuda').requires_grad_()
        test_grad = ref_grad.detach().clone().to(device='cuda')
        test_mod = torch.nn.BatchNorm2d(size[1], device='cuda', dtype=torch.float) # mixed
        test_mod.load_state_dict(ref_mod.state_dict())

        # run test

        # force to use Native BatchNorm if test_backend is 'native'
        with torch.backends.cudnn.flags(enabled=False) if test_backend == 'native' else contextlib.nullcontext():
            out = test_mod(test_input)
            out.backward(test_grad)

        # run reference model
        ref_out = ref_mod(ref_input)
        ref_out.backward(ref_grad)

        err_count = self._assertEqualWrapper(out, ref_out, "Output mismatch")   
        err_count = err_count + self._assertEqualWrapper(test_mod.weight.grad, ref_mod.weight.grad, "Weight gradient mismatch")
        err_count = err_count + self._assertEqualWrapper(test_mod.bias.grad, ref_mod.bias.grad, "Bias gradient mismatch")     
        err_count = err_count + self._assertEqualWrapper(test_mod.running_mean, ref_mod.running_mean, "Running mean mismatch")
        err_count = err_count + self._assertEqualWrapper(test_mod.running_var, ref_mod.running_var, "Running variance mismatch")
        err_count = err_count + self._assertEqualWrapper(test_input.grad, ref_input.grad, "Input gradient mismatch")

        if err_count > 0:
            raise AssertionError(f"Test failed with {err_count} errors.")

instantiate_parametrized_tests(TestBatchNormNative)

if __name__ == '__main__':
    run_tests()
```

Native backend test `python test_batchnorm_native.py -v -k test_batchnorm_backend_native`
Can be run on ROCm or CUDA:
```
python test_batchnorm_native.py -v -k test_batchnorm_backend_native

  test_batchnorm_backend_native (__main__.TestBatchNormNative.test_batchnorm_backend_native) ... 

Weight gradient mismatch >>> Tensor-likes are not close!
Mismatched elements: 8 / 8 (100.0%)
Greatest absolute difference: 0.0200042724609375 at index (0,) (up to 1e-05 allowed)
Greatest relative difference: 0.05601941794157028 at index (7,) (up to 1.3e-06 allowed)

Bias gradient mismatch >>> Tensor-likes are not close!
Mismatched elements: 7 / 8 (87.5%)
Greatest absolute difference: 0.009765625 at index (1,) (up to 1e-05 allowed)
Greatest relative difference: 0.003190184012055397 at index (2,) (up to 1.3e-06 allowed)

Input gradient mismatch >>> Tensor-likes are not close!
Mismatched elements: 3 / 128 (2.3%)
Greatest absolute difference: 0.001953125 at index (1, 4, 0, 1) (up to 1e-05 allowed)
Greatest relative difference: 0.036376953125 at index (2, 0, 0, 0) (up to 0.016 allowed)
======================================================================
FAIL [0.187s]: test_batchnorm_backend_native (__main__.TestBatchNormNative.test_batchnorm_backend_native)
```

MIOpen backend test Is avaliable on ROCm 6.4 only from commit https://github.com/pytorch/pytorch/commit/f402eed4d9515dee5bb9b11a2b157de57a5988c1
Command to run `python test_batchnorm_native.py -v -k test_batchnorm_backend_miopen`
```
python test_batchnorm_native.py -v -k test_batchnorm_backend_miopen

test_batchnorm_backend_miopen (__main__.TestBatchNormNative.test_batchnorm_backend_miopen) ... ok (0.188s)
```

### Versions

It does not depend on the processor (intel/amd), GPU (gfx90a/gfx942/gfx1100/V100), Python (3.10/3.12) or Pytorch version (2.5-2.8). I see the difference in gradients on CUDA also (V100)

PR to add 2D/3D batchnorm tests https://github.com/pytorch/pytorch/pull/156498
I'll update this with commit hash when PR is merged

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

**Comment by soulitzer** [comment]
Given that we don't guarantee bitwise equivalence between different implementations and that bfloat16 only has about 3 digits of relative precision, these differences don't look that bad. I'm going to tenatively close this issue, but feel free to reopen if you have reason to believe that 1e-5 is the still the right level tolerance to check.

**Comment by dnikolaev-amd** [comment]
@soulitzer 
Relative difference 0.05 is greater than 3 digits. 
```
Weight gradient mismatch
Greatest relative difference: 0.05601941794157028 at index (7,) (up to 1.3e-06 allowed)
```

**Comment by soulitzer** [comment]
Fair, I guess I meant maybe more like 2-3 actually (since it only has 7 fraction bits). And when you do more than one operation numerical inaccuracies can build up.
